Pause and Unpause

slintes opened the pull request:
**What this PR does / why we need it**:
Added pause and unpause functionality for VMIs

**Special notes for your reviewer**:

Todo:
- [x] VM(?) / VMI status update via condition
- [x] write unit tests
- [x] write more func tests
- [ ] user-guide update

**Release note**:
```release-note
Added pause and unpause functionality for VMIs
```

kbidarkar commented on pkg/virtctl/suspend-resume/suspend.go:
```diff
@@ -0,0 +1,174 @@
+
+func NewSuspendCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "suspend vm|vmi (VM)|(VMI)",
```
I thought we only would run "suspend on VM" and not VMI, as mostly virtctl commands are used mostly with VM and not VMI.

Feels like an exception for SUSPEND/RESUME for VMI, when using VIRTCTL.

I say this as VIRTCTL help mostly talks about "Virtual Machine/VM" and not much of "VMI".

@slintes: I know we discussed this already, but just wanted to add what I thought here and not because I disagree. :)

slintes replied:
Some commands (start/stop/restart) don't make sense on VMI, other do (console/vnc).

Suspend/resume makes more sense on the VMI than VM IMHO. That's why I decided to implement it on the VMI API only.
Just for more comfort, virtctl will work on VMs too. But internally virtctl is looking for the VMI's name and calling the VMI's API.
Nice (IMHO) side effect: the syntax is more aligned with kubectl, since almost all kubectl commands are like `kubectl <verb> <resourceType>[/ ]<resourceName>`

rmohr commented:
> The virt-manager ui is using "Pause the virtual machine" in the tooltip for this functionality.

I think it is difficult. Under the hood this is `suspend` in libvirt. I think I would prefer a consistent naming across the stack, but I don't mind too much at the end.

fabiand commented:
I mainly worry about the user and not so much about internals.

And my concerns layed out above were mainly driven from my user perspective
where this can lead to confusion.

>

davidvossel commented:
> The virt-manager ui is using "Pause the virtual machine" in the tooltip for this functionality.

yes.

pause = freeze it at the hypervizor layer saving ram and resume by restoring that ram picking up exactly where the guest left off.

suspend = acpi call to suspend the guest internally. the equivalent of closing your laptop lid. 

This is consistent terminology. I can site multiple cases where this is the understanding.

[GCE](https://cloud.google.com/sdk/gcloud/reference/alpha/compute/instances/suspend): suspend is an acpi call.
[VirtualBox](https://superuser.com/questions/1364306/what-is-the-difference-between-pause-and-save-state-in-virtualbox): pause is when cpu simply just stops running commands. 
[AWS/EC2](https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-ec2-now-lets-you-pause-and-resume-your-workloads/): pause/resume a vm by saving ram to an EBS volume and then resuming where it left off.

michalskrivanek commented:
> The term "suspend" is pretty tightly coupled to i.e. supending a laptop when the lid of a laptop is closed. And technically tihs involves asking the guest to suspend.

In virtualization world the term Suspend/Resume just refers to the observable behavior, which is the same as closing and opening the laptop. The internals do not matter that much to the user (i.e. whether it's using S3/S4(not supported by QEMU) or "migrate to file")

This PR seems to implement pausing, which is entirely different (and arguably less interesting), as David says the cpu just simply stops running. It still consumes resources.

davidvossel commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1022,14 +1052,14 @@ func (l *LibvirtDomainManager) SyncVMI(vmi *v1.VirtualMachineInstance, useEmulat
 		}
 		logger.Info("Domain started.")
-	} else if cli.IsPaused(domState) {
+	} else if cli.IsPaused(domState) && !l.paused.contains(vmi.UID) {
```
what is the `l.paused` map book keeping protecting here? I'm not entirely sure I understand why the original `if cli.IsPaused()` statement existed to begin with.

slintes replied:
I understood that qemu can pause VMs under some circumstances without the user requesting it. SyncVMI then should try to unpause.
In order to prevent unpausing VMs which were paused by users, I keep track of them in the paused struct.
While writing this I'm wondering if it would be enough to check for the `ReasonPausedUser` state change reason on the domain...?

slintes commented:
> I like where this is going 
> 

thanks for the review!

> I think there are a few of details we should tease apart just to make sure we're comfortable with this long term. The primary thing I'm trying to wrap my head around is the relationship (or potentially lack there of) of the start/stop/restart actions (which take place on the VM) versus the pause/unpause actions (which take place on the VMI)
> 
> Here are some questions that might be worth discussing.
> 
> * does the fact that these actions are not all within the VM API matter?

I decided to implement this on the VMI because imho it makes sense to be able to pause/unpause a VMI directly, because the VMI resource continues to exist. While for start/stop/restart you create/delete/recreate the VMI. Just for convenience I decided to offer pause/unpause on the VM for virtctl too, but it calls the subresource of the VMI. 

> * do we care that pause/unpause is making a round trip call directly into the libvirt api while start/stop/restart are processed by the virt-handler loop?

Declarative vs imperative was part of the initial discussion about this feature on [0], see comments on the virt-api section. We decided to go with imperative for now, without being 100% sure if it really is the better way. I'm open to further discussion, as long as we can come to a conclusion everyone can live with ;)

> * start/stop/restart occur after the api completes making them an async request, while pause/unpause appears to occur during the API request. Is there any concern with respects to blocking of the http call that we should be concerned with?

Fair point. Let me at least check if a timeout on the http call is configured.

> * what should happen when someone calls stop/start/restart on a paused vm?

start: didn't discuss with QE, but I guess nothing should happen (need to add a test)
stop: accodring to QE, VMI will be stopped the same way as if it was unpaused (there is a test already)
restart: good question. I'd say VMI should be restarted, not sure if it should be suspended immediately. @kbidarkar wdyt? Needs a test either way.

> * if we implement suspend/resume later on, where will that fit into all of this.

good question...

@rmohr thoughts?

[0] https://docs.google.com/document/d/18kGff9CnLi1YNNb7WgGCK6Xyj9KYGGHqFAhl7vO2eWM/edit#

slintes commented:
> > * start/stop/restart occur after the api completes making them an async request, while pause/unpause appears to occur during the API request. Is there any concern with respects to blocking of the http call that we should be concerned with?
> 
> Fair point. Let me at least check if a timeout on the http call is configured.

There is a 10 second timeout on the call from virt-api to virt-handler. Good enough?

> 
> > * what should happen when someone calls stop/start/restart on a paused vm?
> 
> start: didn't discuss with QE, but I guess nothing should happen (need to add a test)

added a test that checks that a paused VM can not be started and expects "VM is already running" error

> stop: according to QE, VMI will be stopped the same way as if it was unpaused (there is a test already)
> restart: good question. I'd say VMI should be restarted, not sure if it should be suspended immediately. @kbidarkar wdyt? Needs a test either way.

added a test which checks that the VM is restarted but not paused again after restart. Feel free to discuss

Apart from potential discussion about implementation details, I think I'm ready with this.

slintes commented:
btw, I find the way how pkg/virt-api/rest/subresource.go and client-go/kubecli/handler.go work together for api to handler communication not very nice. Not sure if the latter should even belong into client-go? However, I didn't want to make a bigger refactoring of it in this PR, but would volunteer to (try to :wink: ) improve this in case others agree in a follow up.

rmohr commented:
> btw, I find the way how pkg/virt-api/rest/subresource.go and client-go/kubecli/handler.go work together for api to handler communication not very nice. Not sure if the latter should even belong into client-go? However, I didn't want to make a bigger refactoring of it in this PR, but would volunteer to (try to  ) improve this in case others agree in a follow up.

Sounds good.

rmohr commented:
> > * what should happen when someone calls stop/start/restart on a paused vm?
> 
> start: didn't discuss with QE, but I guess nothing should happen (need to add a test)
> stop: accodring to QE, VMI will be stopped the same way as if it was unpaused (there is a test already)
> restart: good question. I'd say VMI should be restarted, not sure if it should be suspended immediately. @kbidarkar wdyt? Needs a test either way.

Let's also not forget, that if you don't want to drastically want to refactor liveness probe support, you will for now have to block pause operations on VMIs which have liveness probes configured (readiness should work as expected).

> > * if we implement suspend/resume later on, where will that fit into all of this.
> 
> good question...

I think that depends on the details. If this means that it should not consume resources when suspended, then we might go the route that the VMI does not exist as such when suspended (it would be more like a snapshot then). If we keep the API, I guess it depends if we expect that the controller tries repeatedly to reach a suspend state, or if we see it as a one-off action.

 I don't see a direct conflict with pause/unpause, no matter how we would implement that then.

slintes commented:
> Let's also not forget, that if you don't want to drastically want to refactor liveness probe support, you will for now have to block pause operations on VMIs which have liveness probes configured (readiness should work as expected).

oh, good point
but... I am not sure if we should really block pausing in this case? There might be usecases where you want to pause and unpause the VM quicker than the liveness probe failure kicks in?

slintes commented:
/retest

fabiand commented:
/retest

>

slintes commented:
one of the new tests is flaky... investigating :thinking: 

```
[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com][level:component]Pausing A valid VM when paused via virtctl [test_id:3083]should gracefully handle console connection
/root/go/src/kubevirt.io/kubevirt/tests/pausing_test.go:310
Expected
    <string>: Timeout trying to connect to the virtual machine instance
to contain substring
     <string>: VMI is paused
/root/go/src/kubevirt.io/kubevirt/tests/pausing_test.go:322
```
/hold

stu-gott commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -77,31 +81,35 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+
+		pod.Spec.NodeName = "mynode"
+		pod.Status.Phase = k8sv1.PodRunning
+		pod.Status.PodIP = "10.35.1.1"
```
Where did this IP come from?

slintes replied:
it's just a random IP for this fake handler pod. I did not change anything here, see deleted lines below, I just made it reusable.

slintes commented:
> I think there might be value in another functional test: pause and unpause the same VMI multiple times.

good idea, done!
also hopefully fixed the flaky test

/hold cancel

ILpinto commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -901,6 +909,109 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+
+		})
+
+		It("Should fail pausing a not running VMI", func() {
```
missing test_id

ILpinto commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -901,6 +909,109 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+
+		})
+
+		It("Should fail pausing a running but paused VMI", func() {
```
ditto

ILpinto commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -901,6 +909,109 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
+
+		})
+
+		It("Should fail unpausing a not running VMI", func() {
```
ditto

ILpinto commented on pkg/virt-controller/watch/vm_test.go:
```diff
@@ -816,6 +816,47 @@ var _ = Describe("VirtualMachine", func() {
+			controller.Execute()
+		})
+
+		It("should remove paused condition", func() {
```
missing test id

ILpinto commented on pkg/virt-launcher/virtwrap/cmd-server/server_test.go:
```diff
@@ -106,6 +106,20 @@ var _ = Describe("Virt remote commands", func() {
+			Expect(err).ToNot(HaveOccurred())
+		})
+
 		It("should list domains", func() {
```
ditto

ILpinto commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -214,6 +214,101 @@ var _ = Describe("Manager", func() {
 			Expect(err).To(BeNil())
 			Expect(newspec).ToNot(BeNil())
 		})
+		It("should not unpause a paused VirtualMachineInstance", func() {
```
missing test_id

ILpinto commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -214,6 +214,101 @@ var _ = Describe("Manager", func() {
+			Expect(err).To(BeNil())
+			Expect(newspec).ToNot(BeNil())
+		})
+		It("should pause a VirtualMachineInstance", func() {
```
ditto

ILpinto commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -214,6 +214,101 @@ var _ = Describe("Manager", func() {
+			Expect(err).To(BeNil())
+			Expect(newspec).ToNot(BeNil())
+		})
+		It("should not try to pause a paused VirtualMachineInstance", func() {
```
ditto

ILpinto commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -214,6 +214,101 @@ var _ = Describe("Manager", func() {
+			Expect(err).To(BeNil())
+			Expect(newspec).ToNot(BeNil())
+		})
+		It("should unpause a VirtualMachineInstance", func() {
```
ditto

ILpinto commented on pkg/virtctl/pause/pause_test.go:
```diff
@@ -0,0 +1,87 @@
+		Expect(cmd.Execute()).To(BeNil())
+	})
+
+	It("should pause VM", func() {
```
missing test id

slintes commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1085,6 +1115,107 @@ func (l *LibvirtDomainManager) getDomainSpec(dom cli.VirDomain) (*api.DomainSpec
+		return nil, err
+	}
+
+	if domState == libvirt.DOMAIN_RUNNING {
```
I think that will also work. But what advantage would it have? Why not fail early if we know we can't do what we want to do?

vladikr replied:
I'm fine with both approaches :) The only advantage I see with `brute force` here is that we may minimize the number of calls to libvirt and the whole operation will take less time. I think it's important as we are under lock here. However, I'm not sure how much time do we actually save.

slintes commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1085,6 +1115,107 @@ func (l *LibvirtDomainManager) getDomainSpec(dom cli.VirDomain) (*api.DomainSpec
+		logger.Reason(err).Error("Parsing domain XML failed.")
+		return nil, err
+	}
+	return &newSpec, nil
```
Good question... I guess there is no good reason to do this, and I see the domain even isn't used on the caller side. Will remove it, thanks!

slintes commented:
/hold

wait for https://github.com/kubevirt/kubevirt/pull/2842 and sync time after unpausing

slintes commented:
> also hopefully fixed the flaky test

did not work, trying an old fashioned ugly sleep now... did not find a better solution :/

kbidarkar commented on tests/pausing_test.go:
```diff
@@ -0,0 +1,436 @@
+
+			})
+
+			It("should gracefully handle being started again", func() {
```
I think we only started the VM once in this testcase, we need to start the VM here once again.

slintes replied:
The goal is to test what happens when you start a paused VM. And not to start it twice...?

kbidarkar replied:
Ah! ok then.

slintes commented:
still flaky tests... investigating :/

/hold

slintes commented:
Where are the flaky tests when you want them. And why are focused tests ignored :thinking: 

/test pull-kubevirt-e2e-os-3.11.0-crio
/test pull-kubevirt-e2e-k8s-1.14.6
/test pull-kubevirt-e2e-k8s-1.15.1-ceph

slintes commented:
/test pull-kubevirt-e2e-os-3.11.0-crio
/test pull-kubevirt-e2e-k8s-1.14.6
/test pull-kubevirt-e2e-k8s-1.15.1-ceph

slintes commented:
I'm tempted to say that disabling the OKD lane fixed the flaky tests... :thinking: 
Will cleanup and give it another try

slintes commented:
/retest

slintes commented:
/retest

slintes commented:
all green, no flaky tests anymore...

/hold cancel

dhiller commented on pkg/virtctl/root.go:
```diff
@@ -36,11 +37,21 @@ func NewVirtctlCommand() *cobra.Command {
 
 	// used in cobra templates to display either `kubectl virt` or `virtctl`
-	cobra.AddTemplateFunc("ProgramName", func() string { return programName })
+	cobra.AddTemplateFunc(
```
@slintes I made a few probes on the compiled binary regarding the help templating, all probes still worked as expected.

slintes replied:
perfect, thanks!

slintes commented on pkg/virtctl/pause/pause.go:
```diff
@@ -0,0 +1,174 @@
+
+func NewChildCommands(parentCommand string, clientConfig clientcmd.ClientConfig) []*cobra.Command {
+	commands := make([]*cobra.Command, 0)
+	commands = append(commands, NewChildCommand(parentCommand, COMMAND_VM_SHORT, "(VM)", clientConfig))
```
not sure if I understand, I am following the `kubectl <verb> <resource> <name>` like e.g. `kubectl expose` does as well?
``` 
$ virtctl pause
Pause a virtual machine

Usage:
virtctl pause [command]

Examples:
# Pause a virtual machine called 'myvm':
virtctl pause vm|vmi myvm

Available Commands:
virtualmachine         Pause a virtual machine.
virtualmachineinstance Pause a virtual machine.
vm                     Pause a virtual machine.
vmi                    Pause a virtual machine.

Flags:
-h, --help   help for pause

Use "virtctl options" for a list of global command-line options (applies to all commands).

$ virtctl pause vmi
accepts 1 arg(s), received 0

$ virtctl pause vmi --help
Pause a virtual machine.

Usage:
virtctl pause vmi (VMI) [flags]

Examples:
# Pause a virtual machine called 'myvm':
virtctl pause vmi myvm

Flags:
-h, --help   help for vmi

Use "virtctl options" for a list of global command-line options (applies to all commands).
```

rmohr replied:
What I mean is that in `kubectl` we don't have sub-commands for every individual resource type:

```bash
$ kubectl expose --help
Expose a resource as a new Kubernetes service.

 Looks up a deployment, service, replica set, replication controller or pod by name and uses the selector for that
resource as the selector for a new service on the specified port. A deployment or replica set will be exposed as a
service only if its selector is convertible to a selector that service supports, i.e. when the selector contains only
the matchLabels component. Note that if no port is specified via --port and the exposed resource has multiple ports, all
will be re-used by the new service. Also if no labels are specified, the new service will re-use the labels from the
resource it exposes.

 Possible resources include (case insensitive):

 pod (po), service (svc), replicationcontroller (rc), deployment (deploy), replicaset (rs)

Examples:
  # Create a service for a replicated nginx, which serves on port 80 and connects to the containers on port 8000.
  kubectl expose rc nginx --port=80 --target-port=8000
  
  # Create a service for a replication controller identified by type and name specified in "nginx-controller.yaml",
which serves on port 80 and connects to the containers on port 8000.
  kubectl expose -f nginx-controller.yaml --port=80 --target-port=8000
  
  # Create a service for a pod valid-pod, which serves on port 444 with the name "frontend"
  kubectl expose pod valid-pod --port=444 --name=frontend
```

slintes replied:
got it, fixed

slintes commented on pkg/controller/conditions.go:
```diff
@@ -46,6 +46,9 @@ func (d *VirtualMachineConditionManager) CheckFailure(vmi *v1.VirtualMachineInst
 }
 
 func (d *VirtualMachineConditionManager) HasCondition(vmi *v1.VirtualMachineInstance, cond v1.VirtualMachineInstanceConditionType) bool {
+	if vmi == nil {
```
in the just pushed, refactored and extended conditions_test.go :wink:

slintes commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -572,6 +616,48 @@ func (app *SubresourceAPIApp) StopVMRequestHandler(request *restful.Request, res
+
+	validate := func(vmi *v1.VirtualMachineInstance) (error, int) {
+		if vmi == nil || vmi.Status.Phase != v1.Running {
+			return fmt.Errorf("VMI is not running"), http.StatusForbidden
```
fair point, changed

slintes commented:
/retest

slintes commented:
/retest

rmohr commented:
/lgtm
/approve

I think that in the case of queued actions on the VM (stop/restart), we are missing some checks, but that can be sorted out in an extra PR.

slintes commented:
/retest

slintes commented:
/retest

slintes commented:
/retest

slintes commented:
/retest

slintes commented:
/retest

slintes commented:
1 unrelated last test failure only
`Expose Expose UDP service on a VMI Expose ClusterIP UDP service [test_id:1535]Should expose a ClusterIP service on a VMI and connect to it`

/override pull-kubevirt-e2e-k8s-1.15.1-ceph

slintes commented:
/cherrypick release-0.23

haslersn commented:
> > * what should happen when someone calls stop/start/restart on a paused vm?
> 
> start: didn't discuss with QE, but I guess nothing should happen (need to add a test)
> stop: accodring to QE, VMI will be stopped the same way as if it was unpaused (there is a test already)
> restart: good question. I'd say VMI should be restarted, not sure if it should be suspended immediately. @kbidarkar wdyt? Needs a test either way.

If you want a paused VM to be immediately paused after a restart, then it does not make sense that "pause" is a command on the VMI. That only makes sense if "pause" is a command on the VM.
